### ISSUE TITLE
QtInteractor inheritance change

### ISSUE THREAD

**Issue by banesullivan** [pyvista/pyvista#600]
#548  broke cell picking on the background plotter. Run the following code on commits 38a8543169420c47e1a675523a636594ebedcd89 (pre-#548 break) then afb96961534561f1d3b7118fb8eb48f5e836cb2b (the merge of #548) and see that there's a segfault when making a selection with cell picking.

@GuillaumeFavelier, any idea what caused this?

```py
import pyvista as pv
mesh = pv.Sphere()

p = pv.BackgroundPlotter()
p.add_mesh(mesh)
p.enable_cell_picking()
```

**Comment by banesullivan** [comment]
this just shipped on `v0.23.2` 🙃

**Comment by banesullivan** [comment]
It's about time we figure out #435

**Comment by banesullivan** [comment]
So. We need to fix this asap and send `0.23.3` out there immediately

RFC @pyvista/developers: should we not merge https://github.com/conda-forge/pyvista-feedstock/pull/20, delete 0.23.2 from PyPI, and let `0.23.2` be a skipped version?

**Comment by banesullivan** [comment]
After chatting on Slack with @akaszynski, this is a big enough bug to warrant removing `0.23.2` from PyPI and skipping that version.

Time to slow down and re-evaluate everything before we push a new release out there.

**Comment by banesullivan** [comment]
Also, I am seeing a significant lag introduced by #548 when rotating the camera. Using the same two commits (pre and post #548) you can see the difference (would share screen captures but GIFs make it hard to tell)

**Comment by GuillaumeFavelier** [comment]
Unfortunately I can't reproduce the segfault but I do reproduce the lag during rotation and from my investigations it's related to `auto_update`:

https://github.com/pyvista/pyvista/blob/38a8543169420c47e1a675523a636594ebedcd89/pyvista/plotting/qt_plotting.py#L420-L424

@banesullivan can you try the following and see if you still have the same behaviour:

```py
import pyvista as pv
mesh = pv.Sphere()

p = pv.BackgroundPlotter(auto_update=False)
p.add_mesh(mesh)
p.enable_cell_picking()
```

**Comment by GuillaumeFavelier** [comment]
I ported this part of the code in #548 thinking that the `update()` function was `BasePlotter.update()`.

I checked on 38a8543 and `BasePlotter.update()` is not called when I expect, what is the correct version of `update()` to call then?
After #548 though, I can confirm that `BasePlotter.update()` is called when the window is modified which represents a lot of consecutive calls especially during mouse interaction (rotation, ...). That explains the lag.

**Comment by banesullivan** [comment]
@akaszynski and @GuillaumeFavelier, I cannot figure out how to keep the changes in #548 and keep backward compatibility for code like that in https://github.com/pyvista/pyvista-support/issues/86 (which is very similar to a few downstream, closed-source projects). The problem is outlined in https://github.com/pyvista/pyvista/pull/603#issuecomment-584469529. Basically we used to be able to add the `QtInteractor` class as a Qt widget but not anymore... should we fix this or break changes and let as many people know as we can? 

It will probably be a headache for all the closed source projects using PyVista with PyQt.... perhaps this is a good moment to justify expanding [`pyvista-gui`](https://github.com/pyvista/pyvista-gui) and have all those projects build off of that API

@math-artist, I believe this also affects your project. How do you feel about this?

**Comment by akaszynski** [comment]
I feel having `pyvista` work for jupyter/`ipython` is bigger than its compatibility for closed source projects.  It's a necessary change, and I was a bit lazy with implementing some of the code (i.e. `_render`) and I'm glad that @GuillaumeFavelier came along and cleaned it up.  I'll just have to update some of my downstream code.

I agree with expanding `pyvista-gui` to encompass the changes.

Also, @banesullivan, I have no idea how you keep up with all of this.  I'm still bogged down with unit testing on Azure...

**Comment by GuillaumeFavelier** [comment]
>  I cannot figure out how to keep the changes in #548 and keep backward compatibility for code like that in pyvista/pyvista-support#86 (which is very similar to a few downstream, closed-source projects).

I think I can help with that. The main difference with the current design of `QtInteractor` is that it does not inherit directly from `QVTKRenderWindowInteractor`. Instead it inherits from the adapter class which uses composition: this class holds an `interactor` attribute which is the actual `QVTKRenderWindowInteractor` (simplified design available in https://github.com/pyvista/pyvista/pull/548#issuecomment-576698170). Because of this modification, some changes are indeed necessary in classes that uses `QtInterator`, it's not a `QtWidget` anymore.

I looked at the code snippet in https://github.com/pyvista/pyvista-support/issues/86#issue-538049094 and this is what I would do:

```diff
        # add the pyvista interactor object
        self.vtk_widget = pv.QtInteractor(self.verticalLayoutWidget)
-       self.verticalLayout.addWidget(self.vtk_widget)
+       self.verticalLayout.addWidget(self.vtk_widget.interactor)
+       self.signal_close.connect(self.vtk_widget.interactor.close)
```

where `signal_close` is a signal sent when the main window is closed (example [here](https://github.com/pyvista/pyvista/blob/master/pyvista/plotting/qt_plotting.py#L815-L827)).

An example demonstrating that it's possible is our `TestWindow`: https://github.com/pyvista/pyvista/blob/master/tests/test_qt_plotting.py#L23-L52

*If those changes are not tolerated (too much effort for downstream project), a solution woud be to use an adapter over the `BasePlotter` class instead.*

**Comment by math-artist** [comment]
@Bane I am currently using a class inspired by the PyVista class, and I am subclassing `BasePlotter `and `QVTKRenderWindowInteractor` directly. It should not affect my project.

I'd like to jelp but this is a bit over my head for now. Is there a potential performance issue possible by using the current structure?

**Comment by math-artist** [comment]
On a side note, separating the Qwidget from the interactor seems more clean to me.

**Comment by banesullivan** [comment]
At this point, #603 resolves the cell picking issue I originally raised the issue for, but this issue has evolved into more of a discussion around the inheritance change for the `QtInteractor` class.

**Comment by akaszynski** [comment]
Fixed in #619 and #603.
